snarky/src/lib.rs

```diff
@@ -0,0 +1,84 @@
+// Shamelessly copied from Snarky's snark_intf.ml
+
+pub trait SnarkyVar<F>: Clone {
+    // TODO: Add, mul, etc.
+    fn constant(f: F) -> Self;
+}
+
+pub trait SnarkyType<F> {
+    type Var: SnarkyVar<F>;
+    type Constraint;
+
+    type OutOfCircuit;
+    type Auxiliary;
+
+    fn to_field_elements(&self) -> (Vec<Self::Var>, Self::Auxiliary);
+    fn of_field_elements(x: (Vec<Self::Var>, Self::Auxiliary)) -> Self;
+
+    fn value_to_field_elements(x: &Self::OutOfCircuit) -> (Vec<F>, Self::Auxiliary);
+    fn value_of_field_elements(x: (Vec<F>, Self::Auxiliary)) -> Self::OutOfCircuit;
+
+    const SIZE_IN_FIELD_ELEMENTS: usize;
+
+    fn constraint_system_auxiliary() -> Self::Auxiliary;
+    fn check<CS>(&self)
+    where
+        CS: SnarkyConstraintSystem<F, Var = Self::Var, Constraint = Self::Constraint>;
+}
+
+pub enum HandleResponse<A, PrevRequest> {
+    Respond(A),
+    Via(PrevRequest),
+}
+
+pub trait SnarkyConstraintSystem<F> {
```

Comment:
that looks like the one we have in circuit-writer. Except I'm not sure what Request/Response is, or why it's generic over constraint/var